src/dynamic_honey_badger/sender_queue.rs

```diff
@@ -0,0 +1,48 @@
+use std::cmp;
+
+use rand::Rand;
+use serde::{de::DeserializeOwned, Serialize};
+
+use super::{Batch, Change, ChangeState, DynamicHoneyBadger, Epoch, Message, Step};
+use sender_queue::SenderQueueFunc;
+use {Contribution, Epoched, NodeIdT};
+
+impl<C, N> SenderQueueFunc<DynamicHoneyBadger<C, N>> for DynamicHoneyBadger<C, N>
+where
+    C: Contribution + Serialize + DeserializeOwned,
+    N: NodeIdT + Serialize + DeserializeOwned + Rand,
+{
+    type Step = Step<C, N>;
+
+    fn max_epoch_with_batch(&self, epoch: Epoch, batch: &Batch<C, N>) -> (Epoch, Option<N>) {
+        (
+            cmp::max(batch.next_epoch, epoch),
+            if let ChangeState::InProgress(Change::Add(ref id, _)) = batch.change {
+                // Register the new node to send broadcast messages to it from now on.
+                Some(id.clone())
+            } else {
+                None
+            },
+        )
+    }
+
+    fn is_accepting_epoch(&self, us: &Message<N>, Epoch((them_era, them_hb_epoch)): Epoch) -> bool {
+        let Epoch((era, hb_epoch)) = us.epoch();
+        era == them_era
+            && (them_hb_epoch <= hb_epoch
+                && hb_epoch <= them_hb_epoch.map(|e| e + self.max_future_epochs as u64))
```

Comment:
Shouldn't we accept it even if `hb_epoch.is_none()` but `them_hb_epoch.is_some()`? If I'm not mistaken, `Option<u64>` is ordered "lexicographically", and `None` is just smaller than every `Some(_)`. Also, `them_hb_epoch` is always `Some(_)`, isn't it?

So I would expect something like:

```rust
if era != them_era {
    return false;
}
match (hb_epoch, them_hb_epoch) {
    (Some(us), Some(them)) => them <= us && us <= them + self.max_future_epochs as u64,
    (None, Some(them)) => true,
    (_, None) => error!("Peer's Honey Badger epoch unknown."),
}
```

~But actually, if it's always `Some(_)`, it would be better to make that clear statically, and just have method parameters `them_era: u64, them_hb_epoch: u64` directly?~ (Edit: Not possible, of course, because `Epoch` is not always the same type… :confused: )